### PR TITLE
ARROW-15006: [Python][Doc] Add five more numpydoc checks to CI

### PR DESCRIPTION
**Pull request by amoeba**
This adds the numpydoc checks GL10, PR04, PR05, RT03, and YD01 to CI and fixes the associated issues in docstrings.

These checks are:

- GL10: reST directives {directives} must be followed by two colons
- PR04: Parameter "{param_name}" has no type
- PR05: Parameter "{param_name}" type should not finish with "."
- RT03: Return value has no description
- YD01: No Yields section found

https://numpydoc.readthedocs.io/en/latest/validation.html

### CODE REVIEW

**Comment by amoeba** [comment]
Hey @jorisvandenbossche, since you reviewed [the last numpydoc PR](https://github.com/apache/arrow/pull/13983), would you want to take a look here? This adds five of the ten checks we discussed ([Jira](https://issues.apache.org/jira/browse/ARROW-15006), [Google Sheet](https://docs.google.com/spreadsheets/d/1NVzQe3MA4Mh-AqsWUmwIZu9r5gB-Lmugz7mge3stYvU)) adding and the diff is relatively straightforward compared to the last one.

File: python/pyarrow/_dataset.pyx [hunk 1]
```diff
@@ -532,12 +532,12 @@ cdef class InMemoryDataset(Dataset):
 
     Parameters
     ----------
-    source : The data for this dataset.
+    source : The data for this dataset
```
Comment (wjones127) [inline python/pyarrow/_dataset.pyx:535]: Why are the periods deleted? Other parameter descriptions, like the ones below, keep theirs.

Comment (wjones127) [reply python/pyarrow/_dataset.pyx:535]: Actually this one makes sense; the description here is in the spot for the datatype, which isn't supposed to have a period. I think this comment needs to be fixed to have a type, then this description should go on the next line.

File: python/pyarrow/_dataset_parquet.pyx [hunk 2]
```diff
@@ -67,11 +67,11 @@ cdef class ParquetFileFormat(FileFormat):
     Parameters
     ----------
     read_options : ParquetReadOptions
-        Read options for the file.
+        Read options for the file
```
Comment (wjones127) [inline python/pyarrow/_dataset_parquet.pyx:70]: This one is puzzling though. Why remove this period?

File: python/pyarrow/_dataset_parquet.pyx [hunk 3]
```diff
@@ -236,9 +236,9 @@ class RowGroupInfo:
 
     Parameters
     ----------
-    id : the group id.
-    metadata : the rowgroup metadata.
-    schema : schema of the rows.
+    id : the group id
```
Comment (wjones127) [inline python/pyarrow/_dataset_parquet.pyx:239]: Similar to before, the place after the colon is for the type, so the description should be moved to the next line I think.

File: python/pyarrow/_dataset.pyx (hunk 1 above)
Comment (amoeba) [reply python/pyarrow/_dataset.pyx:535]: Good eye. Numpydoc catches the first issue (the period) but doesn't catch the ultimate issue which is a poorly-formed docstring. I was focusing too narrowly while fixing these. I don't think we have an automated way of catching this type of issue (a non-type sitting in the position a type would normally go).

File: python/pyarrow/_dataset_parquet.pyx (hunk 2 above)
Comment (amoeba) [reply python/pyarrow/_dataset_parquet.pyx:70]: Not sure what I was doing there. I'll add the period back in.

File: python/pyarrow/_dataset_parquet.pyx (hunk 3 above)
Comment (amoeba) [reply python/pyarrow/_dataset_parquet.pyx:239]: Same response as above too, though I think I can fix this in this PR.

**Comment by amoeba** [comment]
Thanks for the review @wjones127. This is ready for another look.

File: python/pyarrow/ipc.pxi [hunk 4]
```diff
@@ -57,11 +57,11 @@ class WriteStats(_WriteStats):
...
-    num_messages : number of messages.
-    num_record_batches : number of record batches.
-    num_dictionary_batches : number of dictionary batches.
-    num_dictionary_deltas : delta of dictionaries.
-    num_replaced_dictionaries : number of replaced dictionaries.
+    num_messages : number of messages
+    num_record_batches : number of record batches
+    num_dictionary_batches : number of dictionary batches
+    num_dictionary_deltas : delta of dictionaries
+    num_replaced_dictionaries : number of replaced dictionaries
```
Comment (wjones127) [inline python/pyarrow/ipc.pxi:64]:
```suggestion
    num_messages : int
        Number of messages.
    num_record_batches : int
        Number of record batches.
    num_dictionary_batches : int
        Number of dictionary batches.
    num_dictionary_deltas : int
        Delta of dictionaries.
    num_replaced_dictionaries : int
        Number of replaced dictionaries.
```

File: python/pyarrow/ipc.pxi [hunk 5]
```diff
@@ -84,11 +84,11 @@ class ReadStats(_ReadStats):
...
```
Comment (wjones127) [inline python/pyarrow/ipc.pxi:91]:
```suggestion
    num_messages : int
        Number of messages.
    num_record_batches : int
        Number of record batches.
    num_dictionary_batches : int
        Number of dictionary batches.
    num_dictionary_deltas : int
        Delta of dictionaries.
    num_replaced_dictionaries : int
        Number of replaced dictionaries.
```

File: python/pyarrow/ipc.pxi [hunk 6]
```diff
@@ -108,14 +108,14 @@ cdef class IpcReadOptions(_Weakrefable):
     ----------
     ensure_native_endian : bool
```
Comment (wjones127) [inline python/pyarrow/ipc.pxi:109]:
```suggestion
    ensure_native_endian : bool, default True
```

File: python/pyarrow/ipc.pxi [hunk 7]
```diff
@@ -108,14 +108,14 @@ cdef class IpcReadOptions(_Weakrefable):
...
         Whether to convert incoming data to platform-native endianness.
-        Default is true.
+        Default is true
```
Comment (wjones127) [inline python/pyarrow/ipc.pxi:111]:
```suggestion
```

File: python/pyarrow/parquet/core.py [hunk 8]
```diff
@@ -1309,7 +1313,7 @@ def get_index(self, key):
 
         Parameters
         ----------
-        key : The value for which we want to known the index.
+        key : The value for which we want to known the index
```
Comment (wjones127) [inline python/pyarrow/parquet/core.py:1316]:
```suggestion
        key : str or int
            The value for which we want to known the index.
```

File: python/pyarrow/parquet/core.py [hunk 9]
```diff
@@ -1222,6 +1225,7 @@ def read(self, columns=None, use_threads=True, partitions=None,
         Returns
         -------
         table : pyarrow.Table
+            The pierce as a pyarrow.Table
```
Comment (wjones127) [inline python/pyarrow/parquet/core.py:1228]:
```suggestion
            The piece as a pyarrow.Table.
```

**Review by wjones127 — CHANGES_REQUESTED** [review]
Saw a couple more instances of description in the type spot.

Also not going to correct the periods missing from end of description, since we aren't planning on enforcing PR09 and RT05, but did include that correct in my examples. (I know Joris has asked this of my PRs in the past.)

**Comment by amoeba** [comment]
Thanks @wjones127, those look good. The help is appreciated :)

**Review by wjones127 — APPROVED** [review]

**Pull request merged**